Sources/PT.PM.Common/UstConstantFolder.cs

```diff
@@ -0,0 +1,281 @@
+            object leftValue = leftFold?.Value;
+            object rightValue = rightFold?.Value;
+            BinaryOperatorLiteral op = binaryOperatorExpression.Operator;
+
+            if ((leftValue is string || leftValue is StringBuilder) && rightFold != null)
+            {
+                if (op.BinaryOperator == BinaryOperator.Plus || op.BinaryOperator == BinaryOperator.Concat)
+                {
+                    // Use StringBuilder instead of immutable strings and concatenation
+                    string rightString = rightValue.ToString();
+
+                    StringBuilder leftStringBuilder;
+                    if (leftValue is string leftString)
+                    {
+                        leftStringBuilder = new StringBuilder((leftString.Length + rightString.Length) * 2);
+                        leftStringBuilder.Append(leftString);
+                    }
+                    else
+                    {
+                        leftStringBuilder = (StringBuilder) leftValue;
+                    }
+                    leftStringBuilder.Append(rightString);
+                    
+                    FoldResult result = ProcessBinaryExpression(binaryOperatorExpression, leftFold, rightFold, leftStringBuilder);
+
+                    if (Logger.IsLogDebugs)
+                    {
+                        Logger.LogDebug(
+                            $"Strings {binaryOperatorExpression} concatenated to \"{leftStringBuilder}\" at {binaryOperatorExpression.TextSpan}");
+                    }
+                    
+                    return result;
+                }
+            }
+            else if (leftValue is long leftInt && rightValue is long rightInt)
+            {
+                long resultInt = 0;
+                bool folded = true;
+                try
+                {
+                    checked
+                    {
+                        switch (op.BinaryOperator)
+                        {
+                            case BinaryOperator.Plus:
+                                resultInt = leftInt + rightInt;
+                                break;
+                            case BinaryOperator.Minus:
+                                resultInt = leftInt - rightInt;
+                                break;
+                            case BinaryOperator.Multiply:
+                                resultInt = leftInt * rightInt;
+                                break;
+                            case BinaryOperator.Divide:
+                                if (rightInt == 0)
+                                {
+                                    folded = false;
+                                    break;
+                                }
+                                resultInt = leftInt / rightInt;
+                                break;
+                            case BinaryOperator.Mod:
+                                resultInt = leftInt % rightInt;
```

Comment:
Don't we need to check `rightInt == 0` here too?

Reply:
Yep, you are right. BTW it is a compiler error:

![image](https://user-images.githubusercontent.com/1150330/47642606-23ec9380-db7a-11e8-9be9-d5af6b86f369.png)